src/pointer-analysis/value_set_dereference.cpp

```diff
@@ -398,8 +398,12 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
     // this is also our guard
     result.pointer_guard = dynamic_object(pointer_expr);
 
-    // can't remove here, turn into *p
-    result.value = dereference_exprt{pointer_expr};
+    // TODO should this be object or root_object?
+    // TODO It's unclear whether this is a good approach to take --- it
+    // successfully ensures that every (non-null) dereference points to
+    // something, making the write set computation work correctly, but dynamic
+    // objects do not seem to be intended to be used in this way.
+    result.value = object;
```

Comment:
The problem is the offset, which might not be known. I believe this should be correct when `o.offset().is_constant() == true`. I think in all other cases you largely need to fall back to what the `byte_extract` base does below.